src/debug/telemetry/telemetry.c

```diff
@@ -0,0 +1,143 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
+//
+// Author: Tobiasz Dryjanski <tobiaszx.dryjanski@intel.com>
+
+#include <zephyr/debug/sparse.h>
+
+#include <sof/audio/module_adapter/module/generic.h>
+#include <sof/debug/telemetry/telemetry.h>
+
+#include <adsp_debug_window.h>
+#include <errno.h>
+#include <limits.h>
+#include <mem_window.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/* Systic variables, one set per core */
+static int telemetry_systick_counter[CONFIG_MAX_CORE_COUNT];
+static int telemetry_prev_ccount[CONFIG_MAX_CORE_COUNT];
+static int telemetry_perf_period_sum[CONFIG_MAX_CORE_COUNT];
+static int telemetry_perf_period_cnt[CONFIG_MAX_CORE_COUNT];
+static struct telemetry_perf_queue telemetry_perf_queue[CONFIG_MAX_CORE_COUNT] = {0};
+
+static void telemetry_perf_queue_append(struct telemetry_perf_queue *q, size_t element)
+{
+	if (!q->full) {
+		q->elements[q->index] = element;
+		q->sum += element;
+		q->index++;
+		q->size++;
+		if (q->index >= AVG_PERF_MEAS_DEPTH) {
+			q->index = 0;
+			q->size = AVG_PERF_MEAS_DEPTH;
+			q->full = true;
+		}
+	} else {
+		/* no space, pop tail */
```

Comment:
are actually two indications of the storage being full needed? You have `q->full` and `q->index >= AVG_PERF_MEAS_DEPTH` - aren't they the same? If they are, maybe better to use only one of them, to improve clarity and to avoid any potential issues keeping them synchronised

Reply:
The queue here is circular. After the queue is full it stays full. We delete the oldest data point when a new one is given.

The first `q->index >= AVG_PERF_MEAS_DEPTH` is used to determine if the queue is full.  After that the index becomes the tail of the queue and `q->index >= AVG_PERF_MEAS_DEPTH` is used to check if the tail reached the last element. The tail then circles back to 0.